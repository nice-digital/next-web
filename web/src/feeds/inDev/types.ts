import {
	AreaOfInterest,
	ProductGroup,
	ProductTypeAcronym,
} from "@/feeds/publications/types";

export enum FeedPath {
	AllProjects = "/gidprojects/all",
	InConsultationProjects = "/inconsultationprojects",
	ProjectDetail = "/project/",
}

export enum TopicSelectionReason {
	"Monitor" = "Monitor - awaiting publication of further literature or evidence",
	"Anticipate" = "Anticipate the topic will be of importance to patients, carers, professionals, commissioners and the health of the public to ensure clinical benefit is realised, inequalities in use addressed, and help them make the best use of NHS resources",
	"NotEligible" = "Not eligible for Health Technology Evaluation guidance",
	"FurtherDiscussion" = "Further discussion required at the Topic Selection Oversight Panel (TSOP)",
}

export enum ProjectStatus {
	"Complete" = "Complete",
	"Discontinued" = "Discontinued",
	"Suspended" = "Suspended",
	"Referred" = "Referred",
	"InProgress" = "InProgress",
	"Proposed" = "Proposed",
	"TopicSelection" = "TopicSelection",
	"ImpactedByCOVID19" = "ImpactedByCOVID19",
}

/** Not a definitive list - this may change over time */
export enum ProjectProcess {
	"MTA" = "MTA",
	"IP" = "IP",
	"TAG" = "TAG",
	"MT" = "MT",
	"DT" = "DT",
	"STAPre2018" = "STA pre-2018",
	"STA2018" = "STA 2018",
	"QSD" = "QSD",
	"CG" = "CG",
	"APG" = "APG",
	"SG" = "SG",
	"HST" = "HST",
	"MIB" = "MIB",
	"STAReview" = "STA Review",
	"PHG" = "PHG",
	"MTAReview" = "MTA Review",
	"CDFReview" = "CDF Review",
	"NG" = "NG",
	"APA" = "APA",
	"ES" = "ES",
	"SC" = "SC",
	"COVID19RapidGuideline" = "COVID-19 rapid guideline",
}

export enum ProjectGroup {
	Guideline = ProductGroup.Guideline,
	Guidance = ProductGroup.Guidance,
	Advice = ProductGroup.Advice,
	Standard = ProductGroup.Standard,
	Other = ProductGroup.Other,
}

export type GIDReference = `GID-${Uppercase<string>}${number}`;

export interface AllProjects {
	links: {
		// In theory InDev supports pagination, but in reality the gidprojects/all feed always returns all projects and a page size of 2048
		first: [
			{
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
		last: [
			{
				// This looks like a bug in InDev as Page=0
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
		self: [
			{
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
	};
	embedded: {
		niceIndevIndevelopmentProject: Project[];
	};
}

export interface Project {
	links: {
		self: [
			{
				href: `/project/${GIDReference}`;
			}
		];
	};
	/** e.g. 01000000-0000-001B-0000-000000005402 */
	eTag: string;
	/** Project reference e.g. GID-TAG377 */
	reference: GIDReference;
	/** Acronym (usually 2 or 3 letters) for the type of the project e.g "TA", "IPG" etc */
	projectType: ProductTypeAcronym;
	/** The name of the project type e.g. "Technology appraisal guidance" for TA or "Interventional procedures guidance" for "IPG" */
	productTypeName: string;
	process: ProjectProcess | string;
	/** Project title e.g. "Atrial fibrillation - ximelagatran [ID376]" or "Clinically isolated syndrome - beta interferons and glatiramer acetate [ID109]" */
	title: string;
	status: ProjectStatus;
	projectGroup: ProjectGroup;
	/**
	 * `null` for unpublished projects or an ISO formatted date string like `2020-12-15T00:00:00`
	 */
	publishedDate: string | null;
	/**
	 * An ISO formatted date string like `2020-09-22T13:51:48.8447645`
	 */
	lastModifiedDate: string;
	/**
	 * Null for unpublished projects or an ISO formatted date string like `2020-10-05T12:27:21.5437767`
	 */
	firstGoLiveDate: string | null;
	/** The date the project was created in in dev */
	createdDate: string;
	developedAs: string | null;
	/** Almost always the same as `DevelopedAs` */
	relevantTo: string | null;
	idNumber: `${number}` | null;
	/** E.g. "BMJ Evidence Centre, BMJ Group", "ScHARR" or "Warwick Evidence" etc */
	evidenceAssessmentGroup: string | null;
	/** TODO Always seems to be an empty array */
	areasOfInterestList: [];
}

export interface InConsultationProjects {
	links: {
		self: [
			{
				href: "/inconsultationprojects?Page=0&PageSize=2048";
			}
		];
	};
	// Sometimes there are no consultations running, in which case `_embedded` is not in the feed at all
	embedded?: {
		niceIndevInconsultationProduct: Consultation[];
	};
	eTag: null;
	// Page and size is hard coded here
	page: 0;
	pageSize: 2048;
	totalItems: number;
}

export interface Consultation {
	_links: {
		self: [
			{
				href: `/guidance/${GIDReference}/consultation/html-content${
					| number
					| ""}`;
			}
		];
	};
	eTag: null;
	reference: GIDReference;
	title: string;
	consultationName: string;
	/** ISO formatted date like `2021-06-18T00:00:00` */
	startDate: string;
	/** ISO formatted date like `2021-06-18T00:00:00` */
	endDate: string;
	consultationType: string;
	resourceTitleId: string;
	projectType: ProjectProcess | string;
	productTypeName: string;
	showExpressionOfInterestSubmissionQuestion: boolean;
	developedAs: null;
	relevantTo: null;
	consultationId: number;
	/** E.g. QSD, IP etc */
	process: string;
	hasDocumentsWhichAllowConsultationComments: boolean;
	isCHTE: boolean;
	/** E.g. HSCTeam or CHTETeam */
	allowedRole: string;
	firstConvertedDocumentId: null;
	firstChapterSlugOfFirstConvertedDocument: null;
	partiallyUpdatedProjectReference: null;
	origProjectReference: null;
	areasOfInterestList: [];
}

export type Link = {
	href: string;
};

type ETag = string | null;

export type IndevFile = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	id: string;
	mimeType: string;
	fileName: string;
	length: number;
	hash: string;
	name: string;
	reference: string;
	resourceTitleId: string;
	consultationDocumentId: number;
};

export type IndevConvertedDocument = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	reference: string;
	title: string;
	resourceTitleId: string;
	projectType: string;
	productTypeName: string;
};

export type IndevResource = {
	links: {
		self: [Link];
	};
	embedded?: {
		niceIndevFile?: IndevFile;
		niceIndevConvertedDocument?: IndevConvertedDocument;
	};
	eTag: ETag;
	title: string;
	level: number;
	publishedDate: string;
	externalUrl: string | null;
	showInDocList: boolean;
	textOnly: boolean;
	consultationId: number;
	consultationDocumentId: number;
	isCurrentViewableConsultationMarkup: boolean;
	convertedDocument: boolean;
	supportsComments: boolean;
	supportsQuestions?: boolean;
};

export type IndevFileResource = IndevResource & {
	embedded: {
		niceIndevFile?: IndevFile;
		niceIndevConvertedDocument?: IndevConvertedDocument;
	};
};

export type IndevConsultation = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	reference: string;
	title: string;
	consultationName: string;
	startDate: string;
	endDate: string;
	consultationType: string;
	resourceTitleId: string;
	projectType: string;
	technologyType: string;
	productTypeName: string;
	showExpressionOfInterestSubmissionQuestion: boolean;
	developedAs: string | null;
	relevantTo: string | null;
	consultationId: number;
	process: string;
	hasDocumentsWhichAllowConsultationComments: boolean;
	isCHTE: boolean;
	allowedRole: string;
	firstConvertedDocumentId: string | null;
	firstChapterSlugOfFirstConvertedDocument: string | null;
	partiallyUpdatedProjectReference: string | null;
	origProjectReference: string | null;
	areasOfInterestList: AreaOfInterest[];
	hidden: boolean;
};

export type IndevConsultationPanel = IndevPanel & {
	embedded: {
		niceIndevConsultation: IndevConsultation;
	};
};

export type IndevPanel = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevResourceList: {
			links: {
				self: [Link];
			};
			embedded: {
				niceIndevResource: IndevResource | IndevResource[];
			};
			eTag: ETag;
			hasResources: boolean;
		};
		niceIndevConsultation?: IndevConsultation;
	};
	eTag: ETag;
	title: string;
	panelType: string;
	legacyPanel: boolean;
	originalReference: string | null;
	updateReference: string | null;
	showPanel: boolean;
};

export type IndevTimeline = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
	additionalInfoLabel: string;
	additionalInfo: string;
	hidden: boolean;
};

export type IndevProjectTeam = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
};

export type IndevLegacyStakeholder = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	item: string;
};

export type IndevConsultee = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
};

export type IndevCommentator = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
};

export type IndevEmailEnquiry = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	item: string;
};

export type IndevProcessHomepage = {
	links: {
		self: [Link];
	};
	etag: ETag;
	description: string;
	linkText: string;
} | null;

export type IndevProjectRelatedLink = {
	links: {
		self: [Link];
	};
	etag: ETag;
	column1: string;
	column2: string;
};

export type NiceIndevProjectRelatedLinkList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevProjectRelatedLink:
			| IndevProjectRelatedLink
			| IndevProjectRelatedLink[];
	};
	etag: ETag;
};

export type IndevSchedule = {
	links: {
		self: [Link];
	};
	etag: ETag;
	column1: string;
	column2: string;
	additionalInfoLabel: string;
	additionalInfo: string;
	hidden: boolean;
};

export type NiceIndevProvisionalScheduleList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevProvisionalSchedule: IndevSchedule | IndevSchedule[];
	};
	etag: ETag;
};

export type NiceIndevTimelineList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevTimeline: IndevTimeline | IndevTimeline[];
	};
};

export type NiceIndevProjectTeamList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevProjectTeam: IndevProjectTeam | IndevProjectTeam[];
	};
};

export type NiceIndevLegacyStakeholderList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevLegacyStakeholder:
			| IndevLegacyStakeholder
			| IndevLegacyStakeholder[];
	};
};

export type NiceIndevConsulteeList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevConsultee: IndevConsultee | IndevConsultee[];
	};
};

export type NiceIndevCommentatorList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevCommentator: IndevCommentator | IndevCommentator[];
	};
};

export type NiceIndevEmailEnquiryList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevEmailEnquiry: IndevEmailEnquiry | IndevEmailEnquiry[];
	};
};

export type IndevUpdate = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	updateReference: string;
};

export type NiceIndevFullUpdateList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevFullUpdate: IndevUpdate | IndevUpdate[];
	};
	eTag: ETag;
};

export type NiceIndevPartialUpdateList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevPartialUpdate: IndevUpdate | IndevUpdate[];
	};
	eTag: ETag;
};

export type IndevTopic = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	item: string;
};

export type NiceIndevTopicList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevTopic: IndevTopic | IndevTopic[];
	};
	eTag: ETag;
};

export enum ProjectType {
	"NG" = "NG",
	"CSG" = "CSG",
	"CG" = "CG",
	"DG" = "DG",
	"MPG" = "MPG",
	"ES" = "ES",
	"HST" = "HST",
	"IPG" = "IPG",
	"KTT" = "KTT",
	"LGB" = "LGB",
	"MIB" = "MIB",
	"MTG" = "MTG",
	"PH" = "PH",
	"QS" = "QS",
	"TA" = "TA",
	"PSG" = "PSG",
	"SC" = "SC",
	"SG" = "SG",
	"PIP" = "PIP",
	"APG" = "APG",
	"APA" = "APA",
	"COV" = "COV",
	"COVM" = "COVM",
	"COVT" = "COVT",
	"COA" = "COA",
	"NGC" = "NGC",
	"IND" = "IND",
	"HTE" = "HTE",
	"HUB" = "HUB",
}

export type ProjectDetail = {
	links: {
		niceIndevStakeholderRegistration: Record<string, unknown>[];
		self: [Link];
	};
	embedded: {
		niceIndevFullUpdateList?: NiceIndevFullUpdateList;
		niceIndevPartialUpdateList?: NiceIndevPartialUpdateList;
		niceIndevTopicList?: NiceIndevTopicList;
		niceIndevPanelList: {
			links: {
				self: [Link];
			};
			embedded: {
				niceIndevPanel: IndevPanel | IndevPanel[];
			};
		};
		niceIndevProvisionalScheduleList?: NiceIndevProvisionalScheduleList;
		niceIndevTimelineList?: NiceIndevTimelineList;
		niceIndevProcessHomepage?: IndevProcessHomepage;
		niceIndevProjectTeamList?: NiceIndevProjectTeamList;
		niceIndevProjectRelatedLinkList?: NiceIndevProjectRelatedLinkList;
		niceIndevLegacyStakeholderList?: NiceIndevLegacyStakeholderList;
		niceIndevConsulteeList?: NiceIndevConsulteeList;
		niceIndevCommentatorList?: NiceIndevCommentatorList;
		niceIndevEmailEnquiryList?: NiceIndevEmailEnquiryList;
	};
	eTag: ETag;
	summary: string | null;
	description: string | null;
	referralDate: string | null;
	suspendDiscontinuedReason: string | null;
	suspendDiscontinuedUrl: string | null;
	suspendDiscontinuedUrlText: string | null;
	legacyModel: boolean;
	productReference: string | null;
	evidenceAssessmentGroup: string | null;
	reference: string;
	projectType: ProjectType | null;
	technologyType: string | null;
	productTypeName: string | null;
	process: string;
	alert: string | null;
	title: string;
	content: string | null;
	status: ProjectStatus;
	projectGroup: ProjectGroup;
	publishedDate: string;
	lastModifiedDate: string;
	firstGoLiveDate: string;
	createdDate: string;
	developedAs: string;
	relevantTo: string | null;
	idNumber: string | null;
	areasOfInterestList: [];
	topicSelectionDecision: string;
	topicSelectionReason: string | null;
	topicSelectionDecisionDate: string | null;
	topicSelectionFurtherInfo: string | null;
	indicatorSubTypes: [];
};

export type resourceInPageNavLink = {
	slug: string;
	title: string;
};

export type niceIndevConvertedDocumentChapter = {
	slug: string;
	title: string;
	href: string;
};

export type niceIndevConvertedDocumentSection = {
	slug: string;
	title: string;
};

export type niceIndevConvertedDocument = {
	_links?: {
		self: [
			{
				href: string;
			}
		];
	};
	eTag?: string | null;
	content: string;
	pdfLink: string | null;
	title?: string;
	chapterTitle?: string;
	chapters: niceIndevConvertedDocumentChapter[];
	sections?: niceIndevConvertedDocumentSection[];
};
