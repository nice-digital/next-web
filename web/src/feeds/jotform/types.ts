export type FormID = `${number}`;

export type Duration = `${number}ms`;

export type NotFoundResponse = {
	responseCode: 404;
	message: string;
	content: "";
	duration: Duration;
	info: string;
};

export type NotAuthorizedResponse = {
	responseCode: 401;
	message: string;
	content: "";
	duration: Duration;
	info: string;
};

export type FormStatus = "ENABLED" | "DISABLED";

export type FormType = "LEGACY" | "CARD";

export type FormSuccessResponse = {
	responseCode: 200;
	message: "success";
	content: {
		id: string;
		username: string;
		title: string;
		height: number;
		status: FormStatus;
		/** Format like "2023-03-21 09:27:18" */
		created_at: string;
		updated_at: string;
		last_submission: string;
		new: `${number}`;
		count: `${number}`;
		type: FormType;
		favorite: `${number}`;
		archived: `${number}`;
		url: string;
	};
	duration: Duration;
};

export type GetFormResponse =
	| NotAuthorizedResponse
	| NotFoundResponse
	| FormSuccessResponse;
