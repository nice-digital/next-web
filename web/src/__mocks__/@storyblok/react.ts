export const storyblokInit = (): null => null;

export const apiPlugin = (): null => null;
