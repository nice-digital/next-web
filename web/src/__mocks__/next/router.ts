export const useRouter = jest.fn().mockImplementation(() => {
	throw new Error("Provide a mock of useRouter for testing purposes");
});
