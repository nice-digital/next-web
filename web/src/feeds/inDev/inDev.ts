import { serverRuntimeConfig } from "@/config";

import { getFeedBodyCached, getFeedBodyUnCached } from "..";
import { ErrorResponse } from "../publications/types";

import {
	Project,
	FeedPath,
	AllProjects,
	Consultation,
	InConsultationProjects,
	ProjectDetail,
	IndevPanel,
} from "./types";

export * from "./types";

const cacheKeyPrefix = "inDev",
	{ defaultTTL, longTTL } = serverRuntimeConfig.cache,
	{ origin, apiKey } = serverRuntimeConfig.feeds.inDev;

export const getAllProjects = async (): Promise<Project[]> =>
	await getFeedBodyCached<Project[]>(
		cacheKeyPrefix,
		FeedPath.AllProjects,
		defaultTTL,
		async () =>
			(
				await getFeedBodyUnCached<AllProjects>(
					origin,
					FeedPath.AllProjects,
					apiKey
				)
			).embedded.niceIndevIndevelopmentProject
	);

export const getAllConsultations = async (): Promise<Consultation[]> =>
	await getFeedBodyCached<Consultation[]>(
		cacheKeyPrefix,
		FeedPath.InConsultationProjects,
		defaultTTL,
		async () =>
			(
				await getFeedBodyUnCached<InConsultationProjects>(
					origin,
					FeedPath.InConsultationProjects,
					apiKey
				)
			)?.embedded?.niceIndevInconsultationProduct || []
	);

/**
 * Gets a project detail.
 *
 */

export const getProjectDetail = async (
	inDevReference: string
): Promise<ProjectDetail | ErrorResponse> =>
	//TODO don't cache error response
	await getFeedBodyCached<ProjectDetail | ErrorResponse>(
		cacheKeyPrefix,
		FeedPath.ProjectDetail + inDevReference,
		longTTL,
		async () =>
			await getFeedBodyUnCached<ProjectDetail | ErrorResponse>(
				origin,
				FeedPath.ProjectDetail + inDevReference,
				apiKey
			)
	);
