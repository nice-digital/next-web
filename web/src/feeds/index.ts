import { type Readable } from "stream";

import axios, { type AxiosInstance } from "axios";
import applyCaseMiddleware from "axios-case-converter";
import { camelCase } from "camel-case";

import { cache, getCacheKey } from "@/cache";

export const client: AxiosInstance = applyCaseMiddleware(
	axios.create({
		headers: { accept: "application/json" },
	}),
	{
		caseFunctions: {
			camel: (input, _options) => {
				// Strip out the prefix from all the embedded keys as it creates so much noise
				return camelCase(input.replace("nice.publications:", ""));
			},
		},
	}
);

/**
 * Gets the body of a feed directly from the back end system
 *
 * @param path The path of the feed to request
 * @returns The body of the feed
 */
export const getFeedBodyUnCached = async <TResponse>(
	origin: string,
	path: string,
	apiKey: string,
	acceptHeader = "application/json"
): Promise<TResponse> => {
	const { data } = await client.get<TResponse>(origin + path, {
		headers: {
			"Api-Key": apiKey,
			Accept: acceptHeader,
		},
		validateStatus: (status: number) => {
			// We don't want feed 404 responses to throw an error, so that we can show users a not found page rather than a server error.
			return (status >= 200 && status < 300) || status == 404;
		},
	});

	return data;
};

/**
 * Gets the body of a feed from the cache, if it exists.
 * If the item isn't in cache then it will call the `getUncachedAction` function to get fresh data.
 *
 * @param path The path of the feed endpoint to request
 * @param ttl The TTL (time to live) of the entry
 * @param getUncachedAction Function that gets the uncached feed directly from the back end.
 * @returns The body of the feed
 */
export const getFeedBodyCached = async <T>(
	groupCacheKey: string,
	path: string,
	ttl: number,
	getUncachedAction: () => Promise<T>
): Promise<T> =>
	cache.wrap<T>(getCacheKey(groupCacheKey.toLowerCase(), path.toLowerCase()), getUncachedAction, { ttl });

/**
 * Gets a response stream from a remote API endpoint, usually used for binary files e.g. PDFs (or mobi/epub etc).
 *
 * @param origin The base URL of the endpoint
 * @param path The path to the endpoint e.g. `/feeds/downloads/737585a0-dad7-4a37-875b-b30b09c3fdc3`
 * @param apiKey The API key for authentication against the API
 * @returns A readable response stream
 */
export const getResponseStream = async (
	origin: string,
	path: string,
	apiKey: string
): Promise<Readable> => {
	const { data } = await axios.get<Readable>(origin + path, {
		headers: {
			"Api-Key": apiKey,
		},
		responseType: "stream",
		maxBodyLength: Infinity,
	});
	return data;
};
