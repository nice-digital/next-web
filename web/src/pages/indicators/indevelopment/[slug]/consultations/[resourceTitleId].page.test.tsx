import { render, screen, within } from "@testing-library/react";
import { useRouter } from "next/router";
import { type GetServerSidePropsContext } from "next/types";

import { FeedPath } from "@/feeds/inDev/types";
import { logger } from "@/logger";
import { addDefaultJSONFeedMocks, axiosJSONMock } from "@/test-utils/feeds";

import ConsultationHTMLPage, {
	ConsultationHTMLPageProps,
	getServerSideProps,
	type Params,
} from "./[resourceTitleId].page";

const productRoot = "indicators",
	resourceTitleId = "html-content-2",
	slug = "gid-ta11102",
	resolvedUrl = `/${productRoot}/indevelopment/${slug}`,
	getServerSidePropsContext = {
		params: {
			slug,
		},
		resolvedUrl,
		query: { productRoot },
	} as unknown as GetServerSidePropsContext<{ slug: string }>;

jest.mock("@/logger", () => ({
	logger: { warn: jest.fn() },
}));

type ConsultationHTMLPagePropsContext = GetServerSidePropsContext<{
	slug: string;
	resourceTitleId: string;
}>;

describe("[resourceTitleId].page", () => {
	describe("getServerSideProps", () => {
		it("should return not found when project doesn't exist", async () => {
			const notFoundIdSlug = "gid-abc123";

			axiosJSONMock.reset();
			axiosJSONMock.onGet(new RegExp(FeedPath.ProjectDetail)).reply(404, {
				Message: "Not found",
				StatusCode: "NotFound",
			});
			addDefaultJSONFeedMocks();

			expect(
				await getServerSideProps({
					...getServerSidePropsContext,
					params: {
						slug: notFoundIdSlug,
						resourceTitleId: resourceTitleId,
					},
				})
			).toStrictEqual({
				notFound: true,
			});
		});

		it("should return not found when no consultation matches resourceTitleId in URL", async () => {
			const notFoundResourceTitleId = "non-existent-html-content";
			expect(
				await getServerSideProps({
					...getServerSidePropsContext,
					params: {
						slug: "gid-ind10244",
						resourceTitleId: notFoundResourceTitleId,
					},
				})
			).toStrictEqual({
				notFound: true,
			});
		});

		it("should return not found if the consultation HTML is empty", async () => {
			axiosJSONMock.reset();
			axiosJSONMock
				.onGet(new RegExp("/guidance/GID-IND10243/consultation/html-content"))
				.reply(404, {
					Message: "Not found",
					StatusCode: "NotFound",
				});
			addDefaultJSONFeedMocks();

			expect(
				await getServerSideProps({
					...getServerSidePropsContext,
					params: {
						slug: "gid-ind10243",
						resourceTitleId: "html-content",
					},
				})
			).toStrictEqual({
				notFound: true,
			});

			expect(logger.warn as jest.Mock).toHaveBeenCalledWith(
				"Consultation HTML not found at /guidance/GID-IND10243/consultation/html-content"
			);
		});

		it.todo("should return props for valid consultation");
	});

	describe("ConsultationHTMLPage", () => {
		// const slug = "gid-dg10049",
		const slug = "gid-ta10914",
			resourceTitleId = "html-content-5",
			productRoot = "indicators",
			resolvedUrl = `/${productRoot}/indevelopment/${slug}/consultations/${resourceTitleId}`,
			context: ConsultationHTMLPagePropsContext = {
				params: { slug, resourceTitleId },
				query: {
					productRoot,
				},
				resolvedUrl,
			} as unknown as ConsultationHTMLPagePropsContext;

		let props: ConsultationHTMLPageProps;
		beforeEach(async () => {
			props = (
				(await getServerSideProps(context)) as {
					props: ConsultationHTMLPageProps;
				}
			).props;
			(useRouter as jest.Mock).mockReturnValue({ asPath: resolvedUrl });
		});

		it("should match snapshot for main content", () => {
			render(<ConsultationHTMLPage {...props} />);
			expect(document.body).toMatchSnapshot();
		});
		describe("SEO", () => {
			it.todo("should render page title with consultation name");
		});

		describe("Breadcrumbs", () => {
			it.todo("should render parent breadcrumb to project overview");
			it.todo("should render consultation name as current breadcrumb");
		});

		it("should render the consultation heading", () => {
			render(<ConsultationHTMLPage {...props} />);
			expect(screen.getByRole("heading", { level: 2 })).toHaveTextContent(
				"Draft guidance: 1"
			);
		});
		it("should render the page last updated date in time tag with ISO date time attribute", () => {
			render(<ConsultationHTMLPage {...props} />);
			const updatedMsg = screen.getByText(
				(_c, el) =>
					el?.textContent === "This page was last updated on 22 February 2023"
			);

			const time = within(updatedMsg).getByText("22 February 2023");

			expect(time).toHaveProperty("tagName", "TIME");
			expect(time).toHaveAttribute("datetime", "2023-02-22");
		});

		it("should render consultation HTML string as HTML", () => {
			render(
				<ConsultationHTMLPage
					{...props}
					consultation={{
						html: "<p>hello</p>",
						title: "anything",
					}}
				/>
			);
			expect(screen.getByText("hello")).toHaveProperty("tagName", "P");
		});
	});
});
