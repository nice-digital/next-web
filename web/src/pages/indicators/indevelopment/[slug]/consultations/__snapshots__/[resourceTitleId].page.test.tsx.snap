// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[resourceTitleId].page ConsultationHTMLPage should match snapshot for main content 1`] = `
<body>
  <div>
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        class="breadcrumbs"
      >
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/standards-and-indicators"
          >
            Standards and Indicators
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/standards-and-indicators/indicators"
          >
            Indicators
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/indicators/indevelopment"
          >
            In development
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/indicators/indevelopment/gid-ta10914"
          >
            GID-TA10914
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <span>
            Draft guidance: 1
          </span>
        </li>
      </ol>
    </nav>
    <script
      type="application/ld+json"
    >
      {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/standards-and-indicators",
        "name": "Standards and Indicators"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/standards-and-indicators/indicators",
        "name": "Indicators"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "@id": "/indicators/indevelopment",
        "name": "In development"
      }
    },
    {
      "@type": "ListItem",
      "position": 5,
      "item": {
        "@id": "/indicators/indevelopment/gid-ta10914",
        "name": "GID-TA10914"
      }
    },
    {
      "@type": "ListItem",
      "position": 6,
      "item": {
        "name": "Draft guidance: 1"
      }
    }
  ]
}
    </script>
    <div
      class="page-header"
      id="content-start"
    >
      <h1
        class="page-header__heading page-header__heading--alt"
      >
        Daratumumab with lenalidomide and dexamethasone for untreated multiple myeloma when stem cell transplant is unsuitable  [ID4014]
      </h1>
      <ul
        class="page-header__metadata"
      >
        <li>
          In development GID-TA10914
        </li>
        <li>
          <span>
            Expected publication date: 
          </span>
          <time
            datetime="23 August 2023"
          >
            23 August 2023
          </time>
        </li>
      </ul>
      <p
        class="page-header__cta"
      >
        <a
          href="/consultations/leadinformation"
        >
          Request commenting lead permission
        </a>
      </p>
    </div>
    <nav
      class="horizontal-nav"
    >
      <ul
        class="horizontal-nav__list"
      >
        <li
          class="horizontal-nav__item"
        >
          <a
            aria-current="false"
            class="horizontal-nav__link"
            href="/guidance/indevelopment/gid-ta10914"
          >
            Project information
          </a>
        </li>
        <li
          class="horizontal-nav__item"
        >
          <a
            aria-current="false"
            class="horizontal-nav__link"
            href="/guidance/indevelopment/gid-ta10914/documents"
          >
            Project documents
          </a>
        </li>
        <li
          class="horizontal-nav__item"
        >
          <a
            aria-current="true"
            class="horizontal-nav__link"
            href="/guidance/indevelopment/gid-ta10914/consultations/html-content-5"
          >
            Consultation
          </a>
        </li>
      </ul>
    </nav>
    <h2>
      Draft guidance: 1
    </h2>
    <div>
      <div>
        <p>
          You can now comment on this draft guidance.
        </p>
        
	
        <p>
          <strong>
            How to comment
          </strong>
        </p>
        
	
        <p>
          You should read both the consultation document and the accompanying evidence (see the committee papers) before making any comments.
        </p>
        
	
        <p>
          You must make your comments directly onto the online version of the consultation document. When you go into the document you will find instructions on how to use the commenting software. You can also review and edit your comments before you submit them. 
        </p>
        
	
        <p>
          You can download a PDF version of the consultation document, but 
          <strong>
            you can only make comments on the online version
          </strong>
          .
        </p>
        
	
        <p>
          <strong>
            Consultation document
          </strong>
        </p>
        
	
        <ul>
          
	
          <li>
            <a
              href="https://www.nice.org.uk/consultations/2095/1/recommendations"
            >
              Appraisal consultation document (online commenting)
            </a>
          </li>
          
	
          <li>
            <a
              href="/guidance/GID-TA10914/documents/129"
            >
              Appraisal consultation document (PDF version)
            </a>
          </li>
          
	
        </ul>
        
	
        <p>
          <strong>
            Closing date for comments: 5pm, 14 March 2023.
          </strong>
        </p>
        
	
        <p>
          <strong>
            About the consultation
          </strong>
        </p>
        
	
        <p>
          We have asked some questions about the consultation that we would like you to answer. You will be able to see the questions when you go into the online consultation document. 
        </p>
        
	
        <p>
          We would also like you to think about whether there are any aspects of the recommendations that need particular consideration to make sure we avoid unlawful discrimination against any group of people. For example, on the grounds of:
        </p>
        
	
        <ul>
          
	
          <li>
            race
          </li>
          
	
          <li>
            sex
          </li>
          
	
          <li>
            disability
          </li>
          
	
          <li>
            religion or belief
          </li>
          
	
          <li>
            sexual orientation
          </li>
          
	
          <li>
            age
          </li>
          
	
          <li>
            gender reassignment
          </li>
          
	
          <li>
            pregnancy and maternity.
          </li>
          
	
        </ul>
        
	
        <p>
          <strong>
            Note that this document is not NICE's final guidance on this technology. The recommendations in section 1 may change after consultation.
          </strong>
        </p>
        
	
        <p>
          After the consultation has closed the appraisal committee will meet again to consider the responses received and prepare the final draft guidance.
        </p>
        
	
        <p>
          <strong>
            Find out more
          </strong>
        </p>
        
	
        <ul>
          
	
          <li>
            <u>
              About this topic
            </u>
          </li>
          
	
          <li>
            <a
              href="https://www.nice.org.uk/process/pmg36/chapter/introduction-to-health-technology-evaluation"
            >
              The appraisal process
            </a>
          </li>
          
	
        </ul>
        
	
        <p>
           
        </p>
      </div>
      

    </div>
    <p>
      This page was last updated on
       
      <time
        datetime="2023-02-22"
      >
        22 February 2023
      </time>
    </p>
  </div>
</body>
`;
